### PR TITLE
🔣 Translate chat message

### LINKED ISSUE #2931: Translate chat messages
- Overview issue https://github.com/nextcloud/spreed/issues/9273

---

- Since Nextcloud 26 a core API to translate strings is available in the server: https://github.com/nextcloud/server/blob/a0ecc37d036ace7bd32675f89b046bf0bff1fe89/core/Controller/TranslationApiController.php#L37
- There should be a … Message actions to translate messages.
- A modal opens translating the message to the currently selected user language and allowing in a dropdown to translate to other languages

### PR DESCRIPTION
**Pull request by rapterjet2004**
Fixes #2931 

Signed-off-by: Julius Linus <julius.linus@nextcloud.com>

### 🖼️ Video

https://github.com/nextcloud/talk-android/assets/69230048/b5413c72-026d-4251-bd2f-e73506bbbc84

### 🚧 TODO
- [x] get functionality working
- [x] refactor code to pass code analysis checks
- [ ]  get it reviewed

### 🏁 Checklist

- [ ] ⛑️ Tests (unit and/or integration) are included or not needed
- [x] 🔖 Capability is checked or not needed 
- [x] 🔙 Backport requests are created or not needed: `/backport to stable-xx.x`
- [x] 📅 Milestone is set
- [x] 🌸 PR title is meaningful (if it should be in the changelog: is it meaningful to users?)

### CODE REVIEW

**Comment by AndyScherzinger** [comment]
@rapterjet2004 I pushed some changes and also rebased the branch, hope that doesn't give you any troubles. So the code checks should be happy and the theming is also in place.

However I ran into the following issue on sermo, not sure there is a nice way out of this (bad UX), also cc @nickvergessen 

So it seems the server tells us "I can do translations", however, the list of translations seems empty, the client has no idea about the source language, hence sends the translation call with a target language only, like:
```https://henlo.nextcloud.com/ocs/v2.php/translation/translate?text=please-translate-me&toLanguage=de```

This will be replied with an HTTP400 my guess is because the server can't detect the source language. So you end up with this:
Screenshot_20230517_122117

File: app/src/main/java/com/nextcloud/talk/translate/TranslateActivity.kt [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+
+    private fun enableSpinners(value: Boolean) {
+        binding.fromLanguageSpinner.isEnabled = value
+        binding.toLanguageSpinner.isEnabled = value
+    }
+
+    private fun translate(fromLanguage: String?, toLanguage: String) {
+        val currentUser: User = userManager.currentUser.blockingGet()
+        val credentials: String = ApiUtils.getCredentials(currentUser.username, currentUser.token)
+        val translateURL = currentUser.baseUrl +
```
Comment (mahibi) [inline app/src/main/java/com/nextcloud/talk/translate/TranslateActivity.kt:143]: lets extract the url creation to ApiUtils where all url's are built...

File: app/src/main/java/com/nextcloud/talk/translate/TranslateActivity.kt [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+
+        val bundle = intent.extras
+        binding.originalMessageTextview.text = bundle?.getString(BundleKeys.KEY_TRANSLATE_MESSAGE)
+        text = bundle?.getString(BundleKeys.KEY_TRANSLATE_MESSAGE)
+    }
+
+    private fun getLanguageOptions() {
+        val currentUser: User = userManager.currentUser.blockingGet()
+        val json = JSONArray(CapabilitiesUtilNew.getLanguages(currentUser).toString())
+        Log.i(TAG, "json is: $json")
```
Comment (mahibi) [inline app/src/main/java/com/nextcloud/talk/translate/TranslateActivity.kt:115]: log can be removed

**Comment by AndyScherzinger** [comment]
I'm off for the next hours @rapterjet2004 but did a last push for the time being.

UI now looks like:
Screenshot_20230517_154110
Screenshot_20230517_154210

File: app/src/main/java/com/nextcloud/talk/translate/TranslateActivity.kt (hunk 2 above)
Comment (AndyScherzinger) [reply app/src/main/java/com/nextcloud/talk/translate/TranslateActivity.kt]: fixed via https://github.com/nextcloud/talk-android/pull/3025/commits/b779a01c274919c464a40ec3e71cfc947071bb92

File: app/src/main/java/com/nextcloud/talk/translate/TranslateActivity.kt (hunk 1 above)
Comment (AndyScherzinger) [reply app/src/main/java/com/nextcloud/talk/translate/TranslateActivity.kt]: fixed via https://github.com/nextcloud/talk-android/pull/3025/commits/b779a01c274919c464a40ec3e71cfc947071bb92

**Comment by AndyScherzinger** [comment]
@mahibi addressed all your comments via https://github.com/nextcloud/talk-android/pull/3025/commits/b779a01c274919c464a40ec3e71cfc947071bb92

@rapterjet2004 I think this one is now good to go - follow-up can be screen rotation issue, a rotation due to the Android life-cycle re-triggers the search

**Review by AndyScherzinger — APPROVED** [review]

**Review by mahibi — APPROVED** [review]

**Comment by rapterjet2004** [comment]
@AndyScherzinger Alright, I made the follow up PR here #3030 to fix the rotation issue + some other things

**Pull request merged**